Add 'edit section details' and 'select section' to section view

mvkski opened the pull request:
This diff includes some refactoring to help facilitate the change
-Add 'size' property to DropdownButton
-Create RefreshAfterEditSectionDialog, like EditSectionDialog but refreshes after form submission, instead of closing
-move out common logic between EditSectionDialog and RefreshAfterEditSectionDialog into separate functions
-Change TeacherDashboardHeader to use Redux
-Move TeacherDashboardNavigation from TeacherDashboardHeader into TeacherDashboard, fixing issue with using Route with Redux
-Move code to switch between sections out of SelectSectionDropdown into library function, for use accross components
-Initialize script selection dropdown logic with current script of section, fixing bug with sections having a scriptId but no course_id

Before
![Screenshot from 2020-01-28 14-22-31](https://user-images.githubusercontent.com/58449474/73310719-b6467180-41d9-11ea-882b-e2af2db4d5b3.png)
After
![TeacherDashboardAfter](https://user-images.githubusercontent.com/58449474/72834140-0dc66980-3c3d-11ea-8ebd-5170daf39c32.png)
Select Section Dropdown
![Screenshot from 2020-01-21 10-39-18](https://user-images.githubusercontent.com/58449474/72834229-3a7a8100-3c3d-11ea-8169-307521089ed9.png)
Edit Section Details
![Screenshot from 2020-01-21 10-39-33](https://user-images.githubusercontent.com/58449474/72834271-49613380-3c3d-11ea-9119-9c99e521276d.png)


# Description

<!--
  A summary of the change, including any relevant motivation and context.

  If relevant, include a description both of the existing behavior and of the new behavior.
-->

<!--
  Other aspects to consider. uncomment and add detail for any that seem necessary:
-->

<!-- ### Background -->
<!-- ### Privacy -->
<!-- ### Security -->
<!-- ### Caching -->
<!-- ### Deployment strategy -->
<!-- ### Future work -->

## Links

<!--
  Any relevant links to external resources; ie, specification documents, jira
  items, related PRs, honeybadger errors, etc
-->

- [spec: Course and Unit improvements](https://docs.google.com/document/d/1vJySqyKozUxknkhY1KyQXWzbV-0c1vdnpKvRsAsVrdI/edit#bookmark=id.d7ljt7jjcmqn)
- [jira # 1021](https://codedotorg.atlassian.net/browse/LP-1021)

## Testing story
Unit tests, E2E tests to be supplied after A/B test of feature has shipped
<!--
  Does your change include appropriate tests?

  If so, please describe how the tests included in this PR are sufficient

  If not, please explain why this change does not need to be tested.
-->

# Reviewer Checklist:

- [ ] Tests provide adequate coverage
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

Erin007 commented:
@mvkski can you include before and after screenshots and a link to the jira ticket and spec please?  That will help me have more context for the review. Thank you!

Erin007 commented on apps/src/templates/teacherDashboard/TeacherDashboardHeader.jsx:
```diff
@@ -26,12 +79,54 @@ export default class TeacherDashboardHeader extends React.Component {
+          <div>
+            <h1 style={styles.shortH1}>{this.selectedSection.name}</h1>
+            <div>
+              <span style={styles.sectionPrompt}>Assigned to: </span>
```
we'll want to make this string and "Edit section details" and "Select section" below translatable by putting them in the i18n pipeline if they aren't already there.

mvkski replied:
Will do 👍

mvkski commented:
@Erin007 Addressed your comments! Now using i18n, added screenshots and links. Also fixed the breaking unit test (pending verification by drone) and added another unit test

Erin007 commented:
Great work so far! There's a lot going on here and you've got it working end to end. WAHOO! 

A few high level questions/ideas: 

1.) I think we'll want to put this behind an experiment flag (a url param) so that Amanda et al can turn on that flag in production and try out the changes.  We often do this so that we can merge changes to production without them actually being visible to users. You could resurrect this [flag](https://github.com/code-dot-org/code-dot-org/pull/31071/files) or use something similar. We can use this experiment flag for A/B testing too.  

2.) Hiding work behind a flag also lets us work incrementally and [ship small changes](https://docs.google.com/document/d/1-x-GcC2xWpUymU6YBWHzxmTAlASWQPMfAf8ASBqWlgA/edit) to production. It's totally a personal preference, but there's opportunity here to break this down into smaller chunks or at least smaller commits moving forward :) 

3.) I wonder if we need a separate component for `<ReloadAfterEditSectionDialog/>`? Could we, instead, do something like [SectionAssigner](https://github.com/code-dot-org/code-dot-org/blob/3af43ff52d1a2751b99ec51814ce4dfea2130d66/apps/src/templates/teacherDashboard/SectionAssigner.jsx#L30) where we pass a prop for whether the "Save" button on the edit section dialog should reload the page or not? Kind of like how the dialog knows whether it's editing or for a new section. 

4.) There's a confusing mismatch between how "Course" is used in the UI and how it's used in the codebase. So, when you say "-Initialize script selection dropdown logic with current script of section, fixing bug with sections having a scriptId but no course_id" I want to be sure we're on the same page. There are many cases where a section will have a script_id but not a course_id, because most assignables are scripts not associated with courses.  For example, the UI says "Jigsaw" is a course, but really in the code it's a script.  The only courses in the code are CSD and CSP.  If it's a course, there's an additional dropdown to select the unit, which in the code is a script. 
<img width="997" alt="Screen Shot 2020-01-21 at 3 45 03 PM" src="https://user-images.githubusercontent.com/12300669/72853023-1af74e80-3c65-11ea-9f0e-c3071d827aca.png">

Feel free to reach out if chatting through this is easier or if you have questions!

mvkski commented:
Thanks Erin! I'll address your points in order.

1 & 2) That sounds like a good idea. I'm currently looking into how A/B testing works at code.org, and on doing that  I should have more context on how experiment flags work as well. From the current vantage point, it sounds like there are a few different ways we could A/B test that may/or may not obviate the need for using something like a feature flag/experiment flag - an A/B test configured to show a new feature to 0% of users functions something like an experiment flag. If you have the time,I'd be happy to talk to you about about experiment flags further as I gain context to confirm weather or not we need one, and ask you follow-up questions.

3.) Your proposed approach would work as well and I think would be good readability-wise too. I think the existing approach is about equally readable, and also has some other advantages.

The approach in the PR it has a little bit more code, but the code it does have has fewer branches (no conditional required to check before a reload, fewer parameters to objects and functions). I chose the approach I chose because I foresee future extensions to EditSectionDialog for other purposes for which I believe the pattern I am using will be very good readability/extensibility wise. As an example of an extension - In an earlier iteration of the work in this PR, I had a version of EditSectionDialog that upon hitting the save button dispatched some events to redux to update the new section details onto the page immidiatley with no refresh. At this time, Amanda decided that she would prefer a reload and so we're going with that, but I could see us returning to an "update redux in place" approach in the future for this form, or possibly making some other extension such as making an additional API call of some sort.

4.) I believe we are on the same page. When I was manually testing the diff, I came upon a situation where a section had a script with a scriptId and no associated courseId, which caused the page to be unable to render because of an issue with the script selection dropdown. The dropdown's logic at the time didn't populate itself with the section's script if no students in the class had participated in the script and the script wasn't part of a course. 

I can try to repro this issue again for additional details

mvkski commented:
- Added changes to the diff to release the above features behind a feature flag, for A/B testing
